readme
"Ultimate list index out of range simulator 2014"
  
Started as a way to mash together different python concepts to prevent forgetting basic syntax
I have no idea where this is going.

How to use: run launch.py, follow instructions.

List of files/folders:

+data:                Files containing data for generators
  +inventory:           Inventory files
    -items              List of items and basic stats
    -attr_mod:          List of basic attribute modifiers
    -atk_def_mod:       List of attack/def modifiers
  +player:            Player files
    -races:             List of player races
    -classes:           List of player classes

+docs:                Documentation folder (html files)

+info:                Information files
  -bugs:                List of bugs and weird behaviours to fix
  -todo:                Ideas and stuff
  -notes:               Temporary notes on feature implementation

+player:              Player files
  -save:                Example save file
  -config:              Example configuration file

+source:              Source file folder
  -dungeon.py:          Dungeon class and generator
  -item.py              Item class
  -launch.py:           Main menus and procedures.
  -mob.py:              Generates a dude with a position.
  -parser.py:           Future word parser (Now empty)
  -player.py:           Basic player functionality.

readme:                This file


info/changelog.md
#Changelog

##Current work

Current stable release: 0.1.7

####**v0.2** (Progress so far):
* New features:
* Bugs Fixed:
  * Fixed excessive multiple leveling bug
  * Fixed wrong message when hunger decreases HP
  * Fixed item pricing bug
  * Fixed bought consumables not going to the last 2 belt positions
* Others:
  * Code refactoring
  * Shippable CI integration
  * Unit tests
  * Moved user interaction to single module
  * Nerfed mana traps

##Old releases

####**v0.1.7** (2014-11-21):
* New features:
* Bugs fixed:
  * Fixed bug in credit scroll
  * [0.1.7b] Fixed space appearing randomly before title
* Others:
  * Code rewritten
  * Removed useless exit prompt

####**v0.1.6** (2014-11-07):
* New features: 
* Bugs fixed:
  * Fixed mobs respawning in the dungeon
  * Fixed item enchant displaying post-enchanting name
* Others: 
  * Rewrote code
  * [0.1.6b] Balanced mob and object generation
  * [0.1.6b] Updated parsing algorithms and code
  * [0.1.16c] Removed unnecesary sys imports
  * [0.1.16d] Merged a request and fixed crashes from request

####**v0.1.5** (2014-10-24):
* New features:
* Bugs fixed:
  * Fixed bad call to item generator
  * [0.1.5b] Fixed crash when picking money up
* Others:
  * Money dropped now depends on player level
  * Updated code
  * Launch.py is now executable
  * Adjusted will test rolls
  * [0.1.5b] Changed default controls

####**v0.1.4** (2014-10-17):
* New features:
* Bugs fixed:
  * Fixed issue with credit scrolling
  * Fixed crash when generating items
* Others:
  * Depurated code
  * Fixed and updated credits
  * Updated default controls

####**v0.1.3** (2014-09-13):
* New features:
* Bugs Fixed:
  * Fixed crash while launching
* Others:
  * Fixed non-critical syntax issues


####**v0.1.2** (2014-08-08):
* New features:
  * Removed the 110 stomach limit
  * Added more food types
* Bugs fixed:
  * Fixed crash when eating inocuous food
* Others:
  * Removed schedule file
  * Removed logo in main menu

####**v0.1.1** (2014-08-01):
* New features:
* Bugs fixed:
* Others:
  * Removed spanish help file
  * Corrected some code

####**v0.1.0** (2014-07-25):
* New features:
  * Potions now spawn on dungeons
* Bugs fixed:
  * Fixed crash when using items
  * Fixed crash when picking items up
  * Fixed game starting only on quick mode
  * Fixed footwear not generating
* Others:
  * Balanced armour
  * [0.1.1] Corrected layout
  * [0.1.1b] Removed some files

####**v0.0.18** (2014-07-18):
* New features:
* Bugs fixed:
  * Crash in mob individual test
  * Mobs having negative secondaries
* Others:
  * Potions appear (rarely) as drops
  * NPCs sell food
  * Mobs with a random level can be spawned
  * Added food to item individual testing

####**v0.0.17** (2014-07-11):
* New features:
* Bugs fixed:
  * Crash when falling through trap
* Others:
  * Food appears randomly on dungeons

####**v0.0.16** (2014-07-04):
* New features:
  * New mob: Dead dog
  * [0.0.16b] Hunger
* Bugs Fixed:
  * Fixed attribute display showing wrong numbers
  * [0.0.16b] Fixed weapon bonuses not disappearing when equipping 2Hs
* Others:
  * Default dungeon size changed to 50x50

####**v0.0.15** (2014-06-27):
* New features:
* Bugs fixed:
  * --EMPTY-- being shown as name when drinking potions
  * [0.0.15c] Crash when falling through traps
* Others:
  * Level increases whenever experience cap is reached.
  * Changed attribute display layout
  * [0.0.15b] Restructured classes
  * [0.0.15c] Default dungeon size changed to 70x70

####**v0.0.14** (2014-06-20):
* New features:
  * Stats
  * Achievements
  * Attack rolls for mobs and players
  * Added extra mobs (Snakes and bats)
  * [0.0.14b] Mob IA
* Bugs Fixed:
  * Crash when loading characters
  * Fixed NPCs selling empty items
  * Minimap showing terrain outside dungeon
  * Race and class selection menu showing twice
  * Crash when accessing belt from inventory
  * 1H and 2H weapons being used at the same time
  * Crash when generating random characters
  * [0.0.14c] Inventory not updating with 2H weapons
  * [0.0.14c] Mobs not disappearing after dying
* Others:
  * Willpower tests have to be passed to attack
  * Hitting flying enemies is now harder

####**v0.0.13** (2014-06-13):
* New features:
  * Loading mobs from files
* Bugs fixed:
  * [0.0.13b] Fixed crash when managing mobs
* Others:
  * Balanced zombies

####**v0.0.12** (2014-06-06):
* New features:
  * Custom game over message
  * [0.0.12+b] Individual module testing 
* Bugs fixed:
  * Fixed formatting issues
  * Fixed consumable generation crashes
* Others:
  * Cleaned code

####**v0.0.11** (2014-05-16):
* New features:
  * Tomes and scrolls
* Bugs fixed:
  * Crash when playing a small dungeon after a large one
* Others:
  * Peddlers have at least one potion for sale
  * [0.0.11b] Cleaned code

####**v0.0.10** (2014-05-09):
* New features:
  * Prestige system
  * Quick play
  * Traps
  * Test utilities
  * Willpower tests
  * Intelligence dependant explored map
* Bugs fixed:
  * Look instruction not working randomly
  * Crashes in potion buying menu
  * [0.0.10b] Fixed crashes in tutorial
* Others:
  * Cleaned crawling interface
  * Removed non working code
  * Expanded parser
  * Improved random character generator
  * [0.0.10b] Improved dungeon generator
  * [0.0.10b] Balanced trap rate/damage
  * [0.0.10b] Finished tutorial

####**v0.0.9** (2014-05-02):
* New features:
  * Potions
* Bugs fixed:
  * Save file having lots of empty items
  * Player having random starting attributes
  * Crash when launching from external directory
* Others:
  * Expanded parser
  * [0.0.9b] Added credits
  * [0.0.9b] Potion saving and loading
  * [0.0.9c] Fixed help module

####**v0.0.8** (2014-04-25):
* New features:
  * Combat system
  * [0.0.8b] Windows support
* Bugs fixed:
  * Some item messages not showing
  * HP and HP changing when opening sheet and loading characters
* Others:
  * Increased mob information
  * [0.0.8c] Charisma and vendor relationship influences price
  

source/item.py
#! /usr/bin/env python 
import os, copy, random, sys
sys.path.append(os.path.abspath("../data/"))
from inventory import *

class consumable:
  """
  Potions, spells, items that can be eaten, drunk or consumed.

  Default attributes

  #Type
  type=4
  
  # 0 - Potion
  #   01 - HP potion
  #   02 - MP potion
  #   03 - Recovery potion (HP&MP)
  #   04 - Status potions #TO-DO
  #   00 - Random (except status potions)
  #
  # Potions regenerate a certain amout of HP or MP in a single turn.
  # The amout of points recovered depends on the type of potion. Each potion type has a fixed value that can vary in the -30% / +30% range with random modifiers.
  # Recovery potions recover both HP and MP, and status potions stop damage over time (Poison, fire, bleeding, etc)
  #
  # 1 - Tome of knowledge
  #
  # Tomes of knowledge add one point to an attribute and decrease one point to another attribute.
  #
  # 2 - Attack items
  #
  # Attack items inflict extra damage. They can be direct things like bombs or extra strenght, or they can inflict extra damage over time.
  #
  # 3.- Unidentified potions/food
  #
  # Unidentified consumables can have totally random and unexpected effects. 
  # Vendors can identify them. When this happens, they are converted to a random type 0, 1 or 2 item. (#TO-DO)
  #
  # 4.- Empty 
  #
  # Generates an object without propieties named "--EMPTY--" to be displayed in the crawl screen.
  # This objects can not be consumend and are for display purposes only.

  #General propieties
  name="--EMPTY--"  #Name
  price=0           #Price

  #Potion propieties
  hpr=0             #Maximum HP recovered
  mpr=0             #Maximum MP recovered
  statusr=0         #If 1, recovers status
  subtype=0         #Potion subtype

  #Tome propieties
  intbst=0          #Intelligence modifier
  dexbst=0          #Dexterity modifier
  perbst=0          #Perception modifier
  conbst=0          #Constitution modifier
  wilbst=0          #Willpower modifier 
  chabst=0          #Charisma modifier
  strbst=0          #Strenght modifier

  #Food propieties
  hungrec=0         #Hunger recovery
  chance=0          #Type of food (Good,risky,bad)

  #Attack propieties #TO-DO
  areatype=0        #Type of area affected
  areasize=0        #Size of the area
  damage=0          #Damage caused
  dps=0             #Damage over time
  """

  def __init__(self,newtype=4,subtype=0):
    """
    Class constructor. Creates a consumable item of the specified type. Needs a subtype parameter for the potions, which is ignored in the rest of the items.

    Consumable items can only be bought from vendors. They can't be found in the ground.
    
    Optional parameters:
    Type integer (Defaults to 4 - empty object)
    Subtype for potions (Defaults to 0 - Random)
    """

    self.reset()
    self.subtype=subtype
    self.type=newtype
    self.statusr=0

    #Process item arrays:
    if newtype==0:
      if subtype!=4: # Status potions are not implemented yet
        typelist=["hp_potions","mp_potions","rec_potions"]
        subtype=random.randrange(1,4) if not subtype else subtype
        data=random.choice(eval("consumables."+typelist[subtype-1]))
        self.name=   data[0]
        self.hpr=    data[1]
        self.mpr=    data[2]
        self.price=  data[3]

    if newtype==1:
      data=random.choice(consumables.tomes)
      self.name=      data[0]
      self.strbst=    data[1]
      self.intbst=    data[2]
      self.dexbst=    data[3]
      self.perbst=    data[4]
      self.conbst=    data[5]
      self.wilbst=    data[6]
      self.chabst=    data[7]
      self.price=     data[8]

    if newtype==3:
      data=random.choice(consumables.food)
      self.name=    data[0]
      self.hungrec= int(data[1])
      self.chance=  int(data[2])
      self.price=   int(data[3])

    #Generate empty object
    if newtype==4: self.name="--EMPTY--"

  def reset(self):
    """
    resets an item so all the attributes are set to zero.
    """

    #Type
    self.type=4

    #General propieties
    self.name="--EMPTY--"
    self.price=0

    #Potion propieties
    self.hpr=self.mpr=self.hungrec=self.chance=0

    #Tome propieties
    self.intbst=self.dexbst=self.perbst=self.conbst=self.wilbst=self.chabst=self.strbst=0

    #Attack propieties
    self.areatype=self.areasize=self.damage=self.dps=0

#Item class. Creates and manages items.
class item:
  """
  # Item types:
  #
  #   01 - Head
  #   02 - Face
  #   03 - Neck
  #   04 - Shoulders
  #   05 - Chest
  #   06 - One hand
  #   07 - Two hands
  #   08 - Ring
  #   09 - Belt
  #   10 - Legs
  #   11 - Feet
  #   00 - Empty (Item with all the attributes set to zero)
  #
  # Attack/Defense modifiers and chance table
  #
  #   01 useless:-100:-100  -> 03.00%
  #   02 rusty:-50:-50      -> 04.00%
  #   03 cracked:-25:-25    -> 06.00%
  #   04 bent:-10:-10       -> 07.00%
  #   05 refined:10:10      -> 05.00%
  #   06 masterful:25:25    -> 04.00%
  #   07 perfect:50:50      -> 00.89%
  #   08 celestial:100:100  -> 00.10%
  #   09 universal:200:200  -> 00.01%
  #   10 no modifiers       -> 70.00%
  #
  #

  Default attributes

  # Item attributes
  name="_"      #Item name
  type=00       #Item table. See any item_XX file or the table below
  equip=0       #Equipped. 1=yes, 2=no

  # Item bonuses
  strbonus=0    #Intelligence boost
  intbonus=0    #Dexterity boost
  dexbonus=0    #Perception boost
  perbonus=0    #Constitution boost
  conbonus=0    #Willpower boost
  wilbonus=0    #Charisma boost
  chabonus=0    #Strenght boost

  # Other
  atk=0         #Attack power
  defn=0        #Defense power
  enchantlv=0   #Enchant level
  """

  def __init__(self,typev):
    """
    Item constructor. 

    Takes a random item from the inventory data file and sets the attributes of a new item with them
    Needs an integer (type) to determine the item type (0 is random)
    """

    # generate array with bonus names
    self.bonuses=["strbonus","intbonus","dexbonus","perbonus","conbonus","wilbonus","chabonus"]

    # Initialize instance
    self.reset()
    self.type=typev

    # Generate list of weapon types (Variables in the items module)
    types=["items_%02i"%i for i in range(1,12)]

    #Assign random item type if the passed type is 0
    typev=random.randint(1,11) if not typev else typev

    # Assign random item in that category
    randitem=random.choice(eval("items.%s"%types[typev-1]))

    #Assign the attributes from a random item in the chosen section
    self.name, self.atk, self.defn, self.price=randitem

    #Add attack and defense modifiers and modifier naming.
    randint=random.randint(1,10000)
    if   randint<=7000: randint=9 #no modifiers
    elif randint>7000 and randint<=7001:  randint= 8  #universal
    elif randint>7001 and randint<=7011:  randint= 7  #celestial
    elif randint>7011 and randint<=7100:  randint= 6  #perfect
    elif randint>7100 and randint<=7500:  randint= 5  #masterful
    elif randint>7500 and randint<=8000:  randint= 4  #refined
    elif randint>8000 and randint<=8700:  randint= 3  #bent
    elif randint>8700 and randint<=9300:  randint= 2  #cracked
    elif randint>9300 and randint<=9700:  randint= 1  #rusty
    elif randint>9700 and randint<=10000: randint= 0  #useless

    #Assign the data with the random values
    if randint!=9:
      modifier=atk_def_modifiers.mods[randint]
      self.name=modifier[0]+" "+self.name
      self.atk+=  (self.atk*modifier[1]/100)
      self.defn+= (self.defn*modifier[2]/100)
      self.price+=(self.price*(modifier[1]+modifier[2])/100)

    # Modifying attribute boosts
    # 20.0% chance of one attribute boost
    # 05.0% chance of two attribute boost
    # 01.0% chance of three attribute boost
    # 00.1% chance of four attribute boost

    randint=random.randint(1,1000)
    attboost=0
    if   randint==1: attboost=4
    elif randint>1  and randint<=10:  attboost=3
    elif randint>10 and randint<=60:  attboost=2
    elif randint>60 and randint<=200: attboost=1
    else: pass

    #Initialize bonuses
    for i in self.bonuses: setattr(self,i,0)

    #Randomly assign the bonus points available
    for i in range(attboost):
      randbonus=self.bonuses[random.randint(0,6)]
      temp=getattr(self,randbonus)
      setattr(self,randbonus,temp+1)

    # Choose one name depending on what attribute boost is higher.
    # Only if the total sum of the values is over zero
    # It gets the names from the attr_modifiers module
    bonusvals=[getattr(self,i) for i in self.bonuses]
    if sum(bonusvals):
      for i,v in enumerate(bonusvals):
        if v>sum(bonusvals)-v: 
          self.name=attr_modifiers.mods[self.bonuses[i]][v]+" "+self.name


    #Adjust price after attr boost
    self.price*=(1+sum(bonusvals))

    #Avoid negative prices 
    self.price=0 if self.price<0 else self.price

    #And, if the type selected was 0, set everything to 0 again
    if not typev: self.reset()

  def reset(self):
    """
    Sets all the attributes of the given item object to zero.
    """

    for i in self.bonuses: exec("self.%s=0"%i)
    self.defn=self.atk=0
    self.name=" "
    self.price=0
    self.enchantlv=self.equip=0    

  def enchant(self):
    """
    Enchants the item 

    Permanently adds attribute bonuses and increases either the attack or defense by 1.
    Requires an item 
    Returns 0 if the enchant failed and 1 if it succeeded.
    Items can only be enchanted up to lv10.

    Chances of attribute boosts:
      79.0% chance of one attribute boost
      15.0% chance of two attribute boost
      04.5% chance of three attribute boost
      00.5% chance of four attribute boost
      01.0% chances of the item being destroyed

    Enchanting an item costs the current item price, and doubles its price.
    If the player has no money to pay for the enchant or the item is lv10, enchant() returns a message and passes.

    If the item is destroyed all its attributes are set to 0. 
    Deleting the resetted item from the inventory is done in the player module after calling the enchant() function.
    """

    # Calculate random number
    randint=random.randint(1,1000)
    attboost=0
    if randint<5: attboost=4
    elif randint>51 and randint<=50: attboost=3
    elif randint>50 and randint<=200: attboost=2
    elif randint>200 and randint<=990: attboost=1
    # If item breaks
    else:        
      self.reset()
      return 0

    # If the item has not broken, enchant it
    # Randomly assign the bonus points available
    for i in range(1,attboost+1):
      boosted=random.randint(0,6)
      t=getattr(self,self.bonuses[boosted])
      setattr(self,self.bonuses[boosted],t+1)

    # Double the price of the item
    # Set price first to avoid enchanted item prices to stay at zero
    self.price=1 if self.price<1 else self.price*2
    

    # Increase attack or defense
    if random.choice([0,1]): self.atk+=1
    else: self.defn+=1

    # Increase enchant level
    self.enchantlv+=1

    #Display the enchanting level in the item name
    if    self.name[-1].isalpha(): self.name=self.name+" +1"
    else: self.name=self.name.split('+')[0]+"+"+str(int(self.name.split('+')[1])+1)
    
    return 1

if __name__=="__main__": pass